samples/go/nomdex/nomdex_update.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package main
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/attic-labs/noms/cmd/util"
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/datas"
+	"github.com/attic-labs/noms/go/dataset"
+	"github.com/attic-labs/noms/go/spec"
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/noms/go/util/status"
+	"github.com/attic-labs/noms/go/walk"
+	humanize "github.com/dustin/go-humanize"
+	flag "github.com/juju/gnuflag"
+)
+
+var (
+	inPathArg  = ""
+	outDsArg   = ""
+	relPathArg = ""
+)
+
+var update = &util.Command{
+	Run:       runUpdate,
+	UsageLine: "up -in-path <path> -out-ds <dspath> -by <relativepath>",
+	Short:     "Build/Update an index",
+	Long:      "Traverse all values starting at root and add values found at 'relativePath' to a map found at 'out-ds'\n",
+	Flags:     setupUpdateFlags,
+	Nargs:     0,
+}
+
+func setupUpdateFlags() *flag.FlagSet {
+	flagSet := flag.NewFlagSet("up", flag.ExitOnError)
+	flagSet.StringVar(&inPathArg, "in-path", "", "a value to search for items to index within ")
+	flagSet.StringVar(&outDsArg, "out-ds", "", "name of dataset to save the results to")
+	flagSet.StringVar(&relPathArg, "by", "", "a path relative to all the items in <in-path> to index by")
+	return flagSet
+}
+
+type StreamingSetEntry struct {
+	valChan chan<- types.Value
+	setChan <-chan types.Set
+}
+
+type IndexMap map[types.Value]StreamingSetEntry
+
+type Index struct {
+	m     IndexMap
+	mutex sync.Mutex
+}
+
+func runUpdate(args []string) int {
+	db, rootObject, err := spec.GetPath(inPathArg)
+	d.Chk.NoError(err)
+
+	if rootObject == nil {
+		fmt.Printf("Object not found: %s\n", inPathArg)
+		return 1
+	}
+
+	outDs := dataset.NewDataset(db, outDsArg)
+	relPath, err := types.ParsePath(relPathArg)
+	if printError(err, "Error parsing -by value\n\t") {
+		return 1
+	}
+
+	typeCacheMutex := sync.Mutex{}
+	typeCache := map[*types.Type]bool{}
```

Comment:
Right. An even more compelling case, imo, is something like .foo["geopos"].
I'm still thinking about this one. Smarter solutions aren't trivial. I'm
looking at have a new method on path called MayResolve(v Value) bool, that
examines the type to see if the path can be resolved but it's not trivial.

On Mon, Aug 15, 2016 at 11:21 AM, Aaron Boodman <notifications@github.com>
wrote:

> In samples/go/nomdex/nomdex_update.go
> <https://github.com/attic-labs/noms/pull/2357#discussion_r74807636>:
>
> > +	db, rootObject, err := spec.GetPath(inPathArg)
> > +	d.Chk.NoError(err)
> > +
> > +	if rootObject == nil {
> > +		fmt.Printf("Object not found: %s\n", inPathArg)
> > +		return 1
> > +	}
> > +
> > +	outDs := dataset.NewDataset(db, outDsArg)
> > +	relPath, err := types.ParsePath(relPathArg)
> > +	if printError(err, "Error parsing -by value\n\t") {
> > +		return 1
> > +	}
> > +
> > +	typeCacheMutex := sync.Mutex{}
> > +	typeCache := map[*types.Type]bool{}
>
> I mentioned this in the other review, but you cannot use the typecache
> this simplistically. Consider a path like .foo[42]. This can resolve on
> one instance of a type, but not another.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/attic-labs/noms/pull/2357/files/57cdc7f5276d6c7698d25191b97f9e144c7ac74e#r74807636>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdvwoG2mR1c-frBXUeVeCjIVTZdBJc5ks5qgK4vgaJpZM4JisYZ>
> .
>